src/background_thread.c

```diff
@@ -0,0 +1,453 @@
+#define JEMALLOC_BACKGROUND_THREAD_C_
+#include "jemalloc/internal/jemalloc_preamble.h"
+#include "jemalloc/internal/jemalloc_internal_includes.h"
+
+#include "jemalloc/internal/assert.h"
+
+/******************************************************************************/
+/* Data. */
+
+/* This option should be opt-in only. */
+#define BACKGROUND_THREAD_DEFAULT false
+/* Read-only after initialization. */
+bool opt_background_thread = BACKGROUND_THREAD_DEFAULT;
+
+/* Protects global state during thread creation / termination. */
+malloc_mutex_t background_thread_lock;
+/* Indicates global state.  Atomic because decay reads this w/o locking. */
+atomic_b_t background_thread_enabled_state;
+size_t n_background_threads;
+/* Thread info per-index. */
+background_thread_info_t *background_thread_info;
+
+/******************************************************************************/
+
+#if !defined(JEMALLOC_BACKGROUND_THREAD)
+#define NOT_REACHED { not_reached(); }
+bool background_thread_create(tsd_t *tsd, unsigned arena_ind) NOT_REACHED
+bool background_threads_init(tsd_t *tsd) NOT_REACHED
+bool background_threads_enable(tsd_t *tsd) NOT_REACHED
+bool background_threads_disable(tsd_t *tsd) NOT_REACHED
+void background_thread_prefork(tsdn_t *tsdn) NOT_REACHED
+void background_thread_postfork_parent(tsdn_t *tsdn) NOT_REACHED
+void background_thread_postfork_child(tsdn_t *tsdn) NOT_REACHED
+#undef NOT_REACHED
+#else
+bool
+background_threads_init(tsd_t *tsd) {
+	assert(have_background_thread);
+	assert(narenas_total_get() > 0);
+
+	background_thread_enabled_set(tsd_tsdn(tsd), opt_background_thread);
+	if (malloc_mutex_init(&background_thread_lock,
+	    "background_thread_global",
+	    WITNESS_RANK_BACKGROUND_THREAD_GLOBAL)) {
+		return true;
+	}
+	background_thread_info = (background_thread_info_t *)base_alloc(
+	    tsd_tsdn(tsd), b0get(), ncpus * sizeof(background_thread_info_t),
+	    CACHELINE);
+	if (background_thread_info == NULL) {
+		return true;
+	}
+
+	for (unsigned i = 0 ; i < ncpus; i++) {
+		background_thread_info_t *info = &background_thread_info[i];
+		if (malloc_mutex_init(&info->mutex, "background_thread",
+		    WITNESS_RANK_BACKGROUND_THREAD)) {
+			return true;
+		}
+		if (pthread_cond_init(&info->cond, NULL)) {
+			return true;
+		}
+		info->started = false;
+	}
+
+	return false;
+}
+
+#define BILLION UINT64_C(1000000000)
+/* Threshold for determining when to wake up the background thread. */
+#define BACKGROUND_THREAD_NPAGES_THRESHOLD UINT64_C(1024)
+/* Background thread max sleep time is decay_time right shift by this. */
+#define BACKGROUND_THREAD_MAX_INTERVAL_SHIFT 2
+/* Minimal sleep interval 200 ms. */
+#define BACKGROUND_THREAD_MIN_INTERVAL_NS (BILLION / 5)
```

Comment:
I feel we still should keep a minimal interval to avoid overwhelming wakeups when the decay_time is small, e.g. when we provide sub-second decay_time support, 1 epoch could be very small, e.g. w/ 0.5 second decay time, 2.5ms or 400 times/sec. How about we set the min interval to be 100ms?